File: programmers/c++/여행경로.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

bool dfs(int k, string start, vector<vector<string>>& tickets, vector<string>& answer, vector<bool>& isUsed) {
	answer.push_back(start);

	if (k == tickets.size())
		return true;
	for (int i = 0; i < tickets.size(); i++) {
		if (!isUsed[i] && start == tickets[i][0]) {
			isUsed[i] = true;
			bool result = dfs(k + 1, tickets[i][1], tickets, answer, isUsed);
			if (result)
				return true;
			isUsed[i] = false;
		}
	}
	answer.pop_back();
	return false;
}
vector<string> solution(vector<vector<string>> tickets) {
	vector<string> answer;
	vector<bool> isUsed(tickets.size(), false);

	sort(tickets.begin(), tickets.end());
	dfs(0, "ICN", tickets, answer, isUsed);

	return answer;
}
int main() {
	vector<vector<string>> tickets = { {"ICN", "SFO"}, {"ICN", "ATL"}, {"SFO", "ATL"}, {"ATL", "ICN"}, {"ATL", "SFO"} };
	vector<string> ans = solution(tickets);
	for (auto a : ans) {
		cout << a << " ";
	}
	return 0;
}

File: programmers/고득점_kit_python/bfs_dfs/여행경로.py
#첫번째 풀이
def dfs(k, start, visited, tickets, answer):
    answer[k] = start
    if k == len(tickets):
        return True

    for i, ticket in enumerate(tickets):
        if ticket[0] == start and not visited[i]:
            visited[i] = 1
            result = dfs(k+1, ticket[1], visited, tickets, answer)
            if result:
                return True
            visited[i] = 0
    return False
    
def solution(tickets):
    answer = [""]*(len(tickets)+1)
    visited = [0]*len(tickets)
    tickets.sort()
    dfs(0, "ICN", visited, tickets, answer)
    return answer
